File: backend/app/services/resources.py
"""Student-facing resource reads: queries, mappers, featured/recommended helpers."""

from __future__ import annotations

from sqlalchemy import String, cast, nulls_last, or_
from sqlalchemy.orm import Session, joinedload

from app import schemas
from app.models import Resource, ResourceCategory as ResourceCategoryModel
from app.models import ResourceStatus as ResourceStatusModel
from app.schemas.enums import ResourceCategory, ResourceStatus


def _load_options():
    return [
        joinedload(Resource.last_edited_by),
        joinedload(Resource.submitted_by),
        joinedload(Resource.reviewed_by),
    ]


def _published_query(db: Session):
    return db.query(Resource).options(*_load_options()).filter(
        Resource.status == ResourceStatusModel.published
    )


def _featured_order(query):
    return query.order_by(
        nulls_last(Resource.featured_order.asc()),
        Resource.published_at.desc(),
    )


def to_resource_response(resource: Resource) -> schemas.ResourceResponse:
    submitted_by = None
    if resource.submitted_by_id is not None and resource.submitted_by is not None:
        submitted_by = schemas.ResourceSubmittedBy(
            id=resource.submitted_by.id,
            full_name=resource.submitted_by.full_name,
            email=resource.submitted_by.email,
        )

    return schemas.ResourceResponse(
        id=resource.id,
        slug=resource.slug,
        title=resource.title,
        category=ResourceCategory(resource.category.value),
        description=resource.description,
        read_time=resource.read_time,
        author=resource.author,
        author_role=resource.author_role,
        image=resource.image,
        image_alt=resource.image_alt,
        body=list(resource.body or []),
        status=ResourceStatus(resource.status.value),
        featured=resource.featured,
        featured_order=resource.featured_order,
        published_at=resource.published_at,
        created_at=resource.created_at,
        updated_at=resource.updated_at,
        last_edited_by=(
            resource.last_edited_by.full_name if resource.last_edited_by else None
        ),
        submitted_by=submitted_by,
        submitted_at=resource.submitted_at,
        reviewed_by=(
            resource.reviewed_by.full_name if resource.reviewed_by else None
        ),
        reviewed_at=resource.reviewed_at,
        rejection_reason=resource.rejection_reason,
        views=resource.views,
        saves=resource.saves,
    )


def list_published_resources(
    db: Session,
    *,
    category: ResourceCategory | None = None,
    search: str | None = None,
) -> list[schemas.ResourceResponse]:
    query = _published_query(db)

    if category is not None:
        query = query.filter(
            Resource.category == ResourceCategoryModel(category.value)
        )

    if search and search.strip():
        term = f"%{search.strip()}%"
        query = query.filter(
            or_(
                Resource.title.ilike(term),
                Resource.description.ilike(term),
                cast(Resource.category, String).ilike(term),
            )
        )

    resources = query.order_by(Resource.published_at.desc()).all()
    return [to_resource_response(resource) for resource in resources]


def list_featured_resources(
    db: Session,
    *,
    limit: int = 2,
) -> list[schemas.ResourceResponse]:
    query = _published_query(db).filter(Resource.featured.is_(True))
    resources = _featured_order(query).limit(limit).all()
    return [to_resource_response(resource) for resource in resources]


def list_recommended_resources(db: Session) -> list[schemas.ResourceResponse]:
    limit = 2
    featured = list_featured_resources(db, limit=limit)
    if len(featured) >= limit:
        return featured

    selected_ids = {item.id for item in featured}
    query = _published_query(db).filter(Resource.featured.is_(False))
    if selected_ids:
        query = query.filter(~Resource.id.in_(selected_ids))

    remaining = limit - len(featured)
    backfill = query.order_by(Resource.published_at.desc()).limit(remaining).all()
    return featured + [to_resource_response(resource) for resource in backfill]


def get_published_resource(db: Session, resource_id: str) -> Resource | None:
    return (
        db.query(Resource)
        .options(*_load_options())
        .filter(
            Resource.id == resource_id,
            Resource.status == ResourceStatusModel.published,
        )
        .first()
    )


def get_related_resources(
    db: Session,
    resource_id: str,
    *,
    limit: int = 3,
) -> list[schemas.ResourceResponse]:
    base = get_published_resource(db, resource_id)
    if base is None:
        return []

    same_category = (
        _published_query(db)
        .filter(
            Resource.id != resource_id,
            Resource.category == base.category,
        )
        .order_by(Resource.published_at.desc())
        .limit(limit)
        .all()
    )

    results = list(same_category)
    if len(results) >= limit:
        return [to_resource_response(resource) for resource in results]

    exclude_ids = {resource_id, *(resource.id for resource in results)}
    other = (
        _published_query(db)
        .filter(
            ~Resource.id.in_(exclude_ids),
            Resource.category != base.category,
        )
        .order_by(Resource.published_at.desc())
        .limit(limit - len(results))
        .all()
    )

    combined = results + list(other)
    return [to_resource_response(resource) for resource in combined]


def increment_resource_views(db: Session, resource_id: str) -> tuple[str, int] | None:
    resource = (
        db.query(Resource)
        .filter(
            Resource.id == resource_id,
            Resource.status == ResourceStatusModel.published,
        )
        .with_for_update()
        .first()
    )
    if resource is None:
        return None

    resource.views += 1
    db.flush()
    return resource.id, resource.views

File: frontend/src/api/resources.js
import { apiFetch } from "./client";



export function mapResource(api) {

  return {

    id: api.id,

    slug: api.slug,

    title: api.title,

    category: api.category,

    description: api.description,

    readTime: api.read_time,

    author: api.author,

    authorRole: api.author_role,

    image: api.image,

    imageAlt: api.image_alt,

    body: api.body ?? [],

    status: api.status,

    featured: api.featured,

    featuredOrder: api.featured_order,

    publishedAt: api.published_at,

    createdAt: api.created_at,

    updatedAt: api.updated_at,

    lastEditedBy: api.last_edited_by,

    submittedBy: api.submitted_by

      ? {

          id: api.submitted_by.id,

          fullName: api.submitted_by.full_name,

          email: api.submitted_by.email,

        }

      : null,

    submittedAt: api.submitted_at,

    reviewedBy: api.reviewed_by,

    reviewedAt: api.reviewed_at,

    rejectionReason: api.rejection_reason,

    views: api.views,

    saves: api.saves,

  };

}



export function mapResourceStats(api) {

  return {

    total: api.total,

    published: api.published,

    drafts: api.drafts,

    pendingReview: api.pending_review,

    featured: api.featured,

    archived: api.archived,

  };

}



export function toApiPayload(formPayload, { publish } = {}) {

  const body = (formPayload.body || [])

    .map((paragraph) => paragraph.trim())

    .filter(Boolean);



  const payload = {

    title: formPayload.title,

    category: formPayload.category,

    description: formPayload.description,

    read_time: formPayload.readTime?.trim() || "5 min read",

    author: formPayload.author,

    author_role: formPayload.authorRole,

    image: formPayload.image,

    image_alt: formPayload.imageAlt,

    body,

  };



  if (publish !== undefined) {

    payload.publish = publish;

  }



  return payload;

}



// --- Student reads (Prompt 7) ---



export async function listResources({ category, search } = {}) {

  const params = new URLSearchParams();

  if (category) params.set("category", category);

  if (search?.trim()) params.set("search", search.trim());

  const query = params.toString();

  const data = await apiFetch(`/resources${query ? `?${query}` : ""}`);

  return (data.resources ?? []).map(mapResource);

}



export async function getFeaturedResources(limit = 2) {

  const data = await apiFetch(`/resources/featured?limit=${limit}`);

  return (data.resources ?? []).map(mapResource);

}



export async function getRecommendedResources() {

  const data = await apiFetch("/resources/recommended");

  return (data.resources ?? []).map(mapResource);

}



export async function getResource(resourceId) {

  const data = await apiFetch(`/resources/${resourceId}`);

  return mapResource(data);

}



export async function getRelatedResources(resourceId) {

  const data = await apiFetch(`/resources/${resourceId}/related`);

  return (data.resources ?? []).map(mapResource);

}



export async function recordResourceView(resourceId) {

  return apiFetch(`/resources/${resourceId}/view`, { method: "POST" });

}



// --- Counsellor CMS ---



export async function listMyResources({ status } = {}) {

  const params = new URLSearchParams();

  if (status) params.set("status", status);

  const query = params.toString();

  const data = await apiFetch(`/counsellor/resources${query ? `?${query}` : ""}`);

  return (data.resources ?? []).map(mapResource);

}



export async function getMyResource(id) {

  const data = await apiFetch(`/counsellor/resources/${id}`);

  return mapResource(data);

}



export async function createResource(payload) {

  const data = await apiFetch("/counsellor/resources", {

    method: "POST",

    body: payload,

  });

  return mapResource(data);

}



export async function updateResource(id, payload) {

  const data = await apiFetch(`/counsellor/resources/${id}`, {

    method: "PUT",

    body: payload,

  });

  return mapResource(data);

}



export async function submitResourceForReview(id) {

  return apiFetch(`/counsellor/resources/${id}/submit`, { method: "POST" });

}



// --- Admin CMS ---



export async function listAdminResources({ status, search } = {}) {

  const params = new URLSearchParams();

  if (status) params.set("status", status);

  if (search?.trim()) params.set("search", search.trim());

  const query = params.toString();

  const data = await apiFetch(`/admin/resources${query ? `?${query}` : ""}`);

  return (data.resources ?? []).map(mapResource);

}



export async function getAdminResourceStats() {

  const data = await apiFetch("/admin/resources/stats");

  return mapResourceStats(data);

}



export async function listPendingReviewResources() {

  const data = await apiFetch("/admin/resources/pending-review");

  return (data.resources ?? []).map(mapResource);

}



export async function getAdminResource(id) {

  const data = await apiFetch(`/admin/resources/${id}`);

  return mapResource(data);

}



export async function createAdminResource(payload, { publish = false } = {}) {

  const data = await apiFetch("/admin/resources", {

    method: "POST",

    body: { ...payload, publish },

  });

  return mapResource(data);

}



export async function updateAdminResource(id, payload) {

  const data = await apiFetch(`/admin/resources/${id}`, {

    method: "PUT",

    body: payload,

  });

  return mapResource(data);

}



export async function publishResource(id) {

  return apiFetch(`/admin/resources/${id}/publish`, { method: "POST" });

}



export async function unpublishResource(id) {

  return apiFetch(`/admin/resources/${id}/unpublish`, { method: "POST" });

}



export async function featureResource(id, { featured, featuredOrder } = {}) {

  return apiFetch(`/admin/resources/${id}/feature`, {

    method: "POST",

    body: {

      featured,

      featured_order: featuredOrder ?? null,

    },

  });

}



export async function archiveResource(id) {

  return apiFetch(`/admin/resources/${id}/archive`, { method: "POST" });

}



export async function restoreResource(id) {

  return apiFetch(`/admin/resources/${id}/restore`, { method: "POST" });

}



export async function reviewResource(id, { decision, rejectionReason } = {}) {

  return apiFetch(`/admin/resources/${id}/review`, {

    method: "POST",

    body: {

      decision,

      rejection_reason: rejectionReason ?? null,

    },

  });

}


